main/src/ca/uwaterloo/flix/language/phase/unification/FastBoolUnification.scala

```diff
@@ -0,0 +1,1715 @@
+  }
+
+  /**
+    * The Successive Variable Elimination algorithm.
+    *
+    * Computes the most-general unifier of the given term `t ~ false` where `fvs` is the list of free variables in `t`.
+    *
+    * Eliminates variables one-by-one from the given list `fvs`.
+    *
+    * Throws a [[BoolUnificationException]] if there is no solution.
+    */
+  private def successiveVariableElimination(t: Term, fvs: List[Int]): BoolSubstitution = fvs match {
+    case Nil =>
+      // Determine if f is unsatisfiable when all constants are made into (flexible) variables.
+      if (!satisfiable(flexify(t)))
+        BoolSubstitution.empty
+      else
+        throw BoolUnificationException()
+
+    case x :: xs =>
+      val t0 = BoolSubstitution.singleton(x, Term.False)(t)
+      val t1 = BoolSubstitution.singleton(x, Term.True)(t)
+      val se = successiveVariableElimination(propagateAnd(Term.mkAnd(t0, t1)), xs)
+
+      val f1 = propagateAnd(Term.mkOr(se(t0), Term.mkAnd(Term.Var(x), Term.mkNot(se(t1)))))
+      val st = BoolSubstitution.singleton(x, f1)
+      st ++ se
+  }
+
+  /**
+    * Returns `t` where every constant has been made into a variable.
+    *
+    * Note: We use the constant name as the variable. This assume that constants and variables are disjoint.
+    */
+  private def flexify(t: Term): Term = t match {
+    case Term.True => Term.True
+    case Term.False => Term.False
+    case Term.Cst(c) => Term.Var(c) // We use the constant name as the variable name.
+    case Term.Var(x) => Term.Var(x)
+    case Term.Not(t) => Term.mkNot(flexify(t))
+    case Term.And(csts, vars, rest) =>
+      // Translate every constant into a variable:
+      Term.mkAnd(csts.map(c => Term.Var(c.c)).toList ++ vars ++ rest.map(flexify))
+    case Term.Or(ts) => Term.mkOr(ts.map(flexify))
+  }
+
+  /**
+    * Returns `true` if the given term `t` is satisfiable, i.e. if there is an assignment to its free variables that
+    * makes the whole term evaluate to true.
+    *
+    * Note that a constant can never be satisfied because we do not know if it is true or false.
+    *
+    * A smarter implementation would use a full-blown SAT solver, but since this function is rarely called and typically
+    * called with a small term, we use a very naive implementation.
+    */
+  private def satisfiable(t: Term): Boolean = t match {
+    case Term.True => true // A true term is already satisfied.
+    case Term.Var(_) => true // A variable is trivially satisfiable.
+    case Term.False => false // A false term can never be satisfied.
+    case _ => evaluateAll(t, t.freeVars.toList, SortedSet.empty) // Evaluate t on all its valuations.
+  }
+
+  /**
+    * Evaluates the given term `t` under all valuations of the free variables `fvs` where `trueVars` are assumed to be true.
+    *
+    * For each variable `x`, the function recursively explores both when `x` is true and when `x` is false.
```

Comment:
It's unclear what the return value indicates

Reply:
Added a comment.